textsearch/csrc/utils.h

```diff
@@ -0,0 +1,29 @@
+// textsearch/csrc/utils.h
+//
+// Copyright (c)  2023  Xiaomi Corporation
+
+#ifndef TEXTSEARCH_CSRC_UTILS_H_
+#define TEXTSEARCH_CSRC_UTILS_H_
+#include <cstdint>
+
+namespace fasttextsearch {
+
+/** This function is copied/modified from
+ * https://github.com/k2-fsa/k2/blob/master/k2/csrc/utils.h#L315
+ *
+ * @param [in] num_elems   The number of elements in the irregular array
+ * @param [in] row_ids   row_ids vector of length num_elems (
+ *                  row_ids[num_elems - 1] + 1 must equal num_rows). Must be
+ *                  non-decreasing.
+ * @param [in] num_rows   Number of rows in the irregular array, must
+ *                  be greater than any element of row_ids
+ * @param [out] row_splits  Row-splits vector that this function
+ *                  writes to, of length num_rows + 1.  row_splits[num_rows]
+ *                  will equal num_elems.
+ */
+void RowIdsToRowSplits(int32_t num_elems, const uint32_t *row_ids,
```

Comment:
I'm not that familiar with this part of k2, could you explain to me what this function is doing / how row splits are going to be used in this project?

Reply:
Let me give you a concret example.

Suppose we have two doc:
  - Doc 0 contains "abc"
  - Doc 1 contains "defgh"

We store the content of all the doc into a single 1-D array. In the above example, the length of the array is
len(abc) + len(defgh) = 3 + 5 = 8.

## The purpose of row id
We need to answer the following question in O(1) time:

  - Which doc is the i-th element in the array from?

The `row_id` in our example is a 1-D array containing
```
0 0 0 1 1 1 1 1
```

- row_id[3] is 1, meaning element 3 in the array, which is `d`, is from doc 1. (note element index starts from 0)

We can know the following extra information from `row_id`:
  - Number of elements in the content array, which is `len(row_id)`, (i.e., 8 in our example)
  - Number of doc, which is `row_id[-1] + 1` (i.e., 2 in our example)
 

# The purpose of row splits

We want to know the following information in O(1) time

- How many elements does the i-th doc contain?
- Where is the first element of the i-th doc in the content array?

We can answer the above two questions by searching `row_id` but not in `O(1)` time.

The purpose of this function `RowIdsToRowSplits` is to compute `row_splits` from `row_ids`.
In our example, `row_ids` is 
```
0 0 0 1 1 1 1 1
```
and the computed `row_splits` would be
```
0 3 5
```

- `row_splits[0]` is 0, meaning the first element of doc 0 starts at position 0 in the content array.
- `row_splits[1]` is 3, meaning the first element of doc 1 starts at position 3 in the content array.

- `row_splits[1] - row_splits[0]` is the number of elements in doc 0
- `row_splits[2] - row_splits[1]` is the number of elements in doc 1
- `len(row_splits) - 1` is the number of doc

---

There is also a function `RowSplitsToRowIds` to compute `row_ids` from `row_splits` in `k2`. But we are not using that in `text_search`.

Reply:
Thanks, this makes sense to me now!